proposals/0698-qualified-literals.rst

```diff
@@ -0,0 +1,565 @@
+
+.. list-table::
+    :align: left
+
+    * - **New expression syntax**
+      - **Desugared expression syntax**
+    * - ``Foo.[]``
+      - ``Foo.buildList (\cons nil -> nil)``
+    * - ``Foo.[x, y]``
+      - ``Foo.buildList (\cons nil -> x `cons` (y `cons` nil))``
+    * - ``Foo.[x ..]``
+      - ``Foo.buildList (Foo.enumFrom x)``
+    * - ``Foo.[x, y ..]``
+      - ``Foo.buildList (Foo.enumFromThen x y)``
+    * - ``Foo.[x .. y]``
+      - ``Foo.buildList (Foo.enumFromTo x y)``
+    * - ``Foo.[x, y .. z]``
+      - ``Foo.buildList (Foo.enumFromThenTo x y z)``
+
+.. list-table::
+    :align: left
+
+    * - **New pattern syntax**
+      - **Desugared pattern syntax**
+    * - ``Foo.[x, _, y]``
+      - ``Foo.FromListCons x (Foo.FromListCons _ (Foo.FromListCons y Foo.FromListNil))``
+    * - ``x Foo.: y``
+      - ``Foo.FromListCons x y``
+
+One might wonder why this doesn't align more closely with the interface of ``-XOverloadedLists``, e.g. ``Foo.fromListN 3 [x, y, z]``. The reason is to avoid the intermediate list, which would need to typecheck as a list. Similar reason for defining new ``enumFrom`` functions instead of reusing Prelude's. See *Section 4.6 Heterogeneous Lists* for a use-case.
+
+We also decide to do ``Foo.buildList`` instead of something like ``Foo.fromList (x `Foo.cons` Foo.nil)`` so that there's one definition to jump to (e.g. with IDE integrations) instead of three.
+
+To use as patterns, the implementor should define ``FromListCons`` and ``FromListNil`` pattern synonyms, typically with the ``COMPLETE`` pragma specified. We choose to do this instead of ``toList -> [x, _, z]`` because that would also disallow heterogeneous lists.
+
+Future work could be done to allow list comprehensions, e.g. ``Foo.[x * 10 | x <- [1..10]]`` => ``[1..10] `Foo.listCompBind` \x -> Foo.listCompReturn (x * 10)``, but that is out of scope of this proposal.
+
+Parser
+~~~~~~
+
+Update `Section 10.5 <https://www.haskell.org/onlinereport/haskell2010/haskellch10.html#x17-18000010.5>`_ of the Haskell 2010 report as follows.
+
+.. code-block:: abnf
+
+  aexp → qvar
+       | ...
+       | modid . integer
+       | modid . ( {-} integer )
+       | modid . ( float )
+       | modid . string
+       | modid . multiLineString
+       | modid . [ exp_1 , ..., exp_k ]
+
+  apat → var [ @ apat ]
+       | ...
+       | modid . integer
+       | modid . ( {-} integer )
+       | modid . ( float )
+       | modid . string
+       | modid . multiLineString
+       | modid . [ pat_1 , ..., pat_k ]
+       | modid . ( pat_1 : ... : pat_k )
+
+
+Proposed Library Change Specification
+-------------------------------------
+
+Template Haskell
+~~~~~~~~~~~~~~~~
+
+We'll add the following constructors, to maintain backwards compatibility:
+
+::
+
+  QualListE ModuleName [Exp]
+
+  QualStringL ModuleName String
+  QualNaturalL ModuleName Natural
+  QualIntegerL ModuleName Integer
+  QualRationalL ModuleName Rational
+
```

Comment:
```suggestion
  data Exp =
  ...
  | QualListE ModuleName [Exp]

  data Lit =
  ...
  | QualStringL ModuleName String
  | QualNaturalL ModuleName Natural
  | QualIntegerL ModuleName Integer
  | QualRationalL ModuleName Rational

```

I think you want to be clear about which datatypes are receiving these  new constructors? I'm not sure about the exact syntax we want to use, but I'm guessing it's something like this